map/bookmark_manager.cpp

```diff
@@ -404,6 +404,59 @@ void BookmarkManager::ResetRecentlyDeletedBookmark()
   m_recentlyDeletedBookmark.reset();
 }
 
+size_t BookmarkManager::GetRecentlyDeletedCategoriesCount() const
+{
+  Platform::FilesList files;
+  Platform::GetFilesByExt(GetTrashDirectory(), kKmlExtension, files);
+  return files.size();
+}
+
+BookmarkManager::KMLDataCollectionPtr BookmarkManager::GetRecentlyDeletedCategories()
+{
+  auto collection = LoadBookmarks(GetTrashDirectory(), kKmlExtension, KmlFileType::Text,
+                                  [](kml::FileData const &)
+                                  {
+    return true;
+  });
+  return collection;
+}
+
+bool BookmarkManager::IsRecentlyDeletedCategory(const std::string &filePath) const
+{
+  Platform::FilesList files;
+  Platform::GetFilesByExt(GetTrashDirectory(), kKmlExtension, files);
+  auto const it = std::find(files.begin(), files.end(), base::FileNameFromFullPath(filePath));
+  return it != files.end();
+}
+
+void BookmarkManager::RecoverRecentlyDeletedCategoriesAtPaths(std::vector<std::string> const & filePaths)
+{
+  CHECK_THREAD_CHECKER(m_threadChecker, ());
+  GetPlatform().RunTask(Platform::Thread::File, [this, filePaths]()
+  {
+    for (auto const & deletedFilePath : filePaths)
+    {
+      CHECK(IsRecentlyDeletedCategory(deletedFilePath), ("The category at path", deletedFilePath, "should be in the trash."));
+      CHECK(Platform::IsFileExistsByFullPath(deletedFilePath), ("File should exist to be recovered.", deletedFilePath));
+      auto recoveredFilePath = GenerateValidAndUniqueFilePathForKML(base::GetNameFromFullPathWithoutExt(deletedFilePath));
+      base::MoveFileX(deletedFilePath, recoveredFilePath);
+      GetPlatform().RunTask(Platform::Thread::Gui, [this, recoveredFilePath]() {
```

Comment:
Why it should be run on UI thread? ReloadBookmark already uses necessary threads internally.

Reply:
@vng is there a check/optimization that if we're already running on the same thread as RunTask is receiving, then the code is run immediately, without putting it into a queue? If no, let's create an issue for it.

Reply:
@biodranik the  Reload adn Load methods cheks the thread first so thay should be called from that main
```cpp
void BookmarkManager::LoadBookmark(std::string const & filePath, bool isTemporaryFile)
{
  CHECK_THREAD_CHECKER(m_threadChecker, ());
  // Defer bookmark loading in case of another asynchronous process.
  if (!m_loadBookmarksFinished || m_asyncLoadingInProgress)
  {
....
}

void BookmarkManager::ReloadBookmark(std::string const & filePath)
{
  CHECK_THREAD_CHECKER(m_threadChecker, ());
...
```

Reply:
This approach looks like a heavy overhead:
1. Run on File thread
2. Move file
3. Run on GUI/main thread
4. Reload bookmarks
5. ReloadBookmarkRoutine
6. Run on File thread
7. Notify on GUI thread

If it's ok to delete files from the main thread, then it should be also ok to move files from the main thread. `RecoverRecentlyDeletedCategoriesAtPaths` is called from the main thread, right? Can step (1) and (3) be removed in this case?